Mathlib/NumberTheory/ArithmeticFunction.lean

```diff
@@ -0,0 +1,1102 @@
+import Mathlib.Algebra.Invertible
+import Mathlib.Data.Nat.Factorization.Basic
+
+/-!
+# Arithmetic Functions and Dirichlet Convolution
+
+This file defines arithmetic functions, which are functions from `ℕ` to a specified type that map 0
+to 0. In the literature, they are often instead defined as functions from `ℕ+`. These arithmetic
+functions are endowed with a multiplication, given by Dirichlet convolution, and pointwise addition,
+to form the Dirichlet ring.
+
+## Main Definitions
+ * `arithmetic_function R` consists of functions `f : ℕ → R` such that `f 0 = 0`.
+ * An arithmetic function `f` `is_multiplicative` when `x.coprime y → f (x * y) = f x * f y`.
+ * The pointwise operations `pmul` and `ppow` differ from the multiplication
+  and power instances on `arithmetic_function R`, which use Dirichlet multiplication.
+ * `ζ` is the arithmetic function such that `ζ x = 1` for `0 < x`.
+ * `σ k` is the arithmetic function such that `σ k x = ∑ y in divisors x, y ^ k` for `0 < x`.
+ * `pow k` is the arithmetic function such that `pow k x = x ^ k` for `0 < x`.
+ * `id` is the identity arithmetic function on `ℕ`.
+ * `ω n` is the number of distinct prime factors of `n`.
+ * `Ω n` is the number of prime factors of `n` counted with multiplicity.
+ * `μ` is the Möbius function (spelled `moebius` in code).
+
+## Main Results
+ * Several forms of Möbius inversion:
+ * `sum_eq_iff_sum_mul_moebius_eq` for functions to a `comm_ring`
+ * `sum_eq_iff_sum_smul_moebius_eq` for functions to an `add_comm_group`
+ * `prod_eq_iff_prod_pow_moebius_eq` for functions to a `comm_group`
+ * `prod_eq_iff_prod_pow_moebius_eq_of_nonzero` for functions to a `comm_group_with_zero`
+
+## Notation
+The arithmetic functions `ζ` and `σ` have Greek letter names, which are localized notation in
+the namespace `arithmetic_function`.
+
+## Tags
+arithmetic functions, dirichlet convolution, divisors
+
+-/
+
+
+open Finset
+
+open BigOperators
+
+namespace Nat
+
+variable (R : Type _)
+
+/-- An arithmetic function is a function from `ℕ` that maps 0 to 0. In the literature, they are
+  often instead defined as functions from `ℕ+`. Multiplication on `arithmetic_functions` is by
+  Dirichlet convolution. -/
+def ArithmeticFunction [Zero R] :=
+  ZeroHom ℕ R
+#align nat.arithmetic_function Nat.ArithmeticFunction
+
+instance ArithmeticFunction.zero [Zero R] : Zero (ArithmeticFunction R) :=
+  inferInstanceAs (Zero (ZeroHom ℕ R))
+
+instance [Zero R] : Inhabited (ArithmeticFunction R) := inferInstanceAs (Inhabited (ZeroHom ℕ R))
+
+variable {R}
+
+namespace ArithmeticFunction
+
+section Zero
+
+variable [Zero R]
+
+instance : FunLike (ArithmeticFunction R) ℕ fun _ ↦ R :=
+  inferInstanceAs (FunLike (ZeroHom ℕ R) ℕ fun _ ↦ R)
+
+@[simp]
+theorem toFun_eq (f : ArithmeticFunction R) : f.toFun = f := by
+  rfl
+#align nat.arithmetic_function.to_fun_eq Nat.ArithmeticFunction.toFun_eq
+
+@[simp]
+theorem eq_toFun (f : ArithmeticFunction R) : f = f.toFun := by
```

Comment:
I've added this by hand, it seems more useful than `toFun_eq` above, but I don't fully understand how coercions work now. In any case without it certain proofs break.

Reply:
I'm surprised this doesn't cause simp to loop with `toFun_eq` also being a `simp` lemma. I think the normal form should be `FunLike.coe`, so we shouldn't have this lemma

Reply:
Since you're version works I am completely happy with it.